proposals/0000-type-variable-scoping.rst

```diff
@@ -0,0 +1,1097 @@
+
+   Note that the pattern signature binds ``b`` to ``Bool``. This is, essentially, a ``let``\ -bound
+   type variable: in the scope of ``b``, ``b`` is synonymous with ``Bool``. Yet the only way
+   to make such a ``b`` is via a pattern (or result, `#228`_) signature. Why force users
+   to use matching instead of binding the variable directly.
+
+#. Doing this helps uphold the EBP_.
+
+Specification
+^^^^^^^^^^^^^
+
+1. Create a new extension ``-XExtendedLet``.
+
+#. With ``-XExtendedLet``, add two new productions for ``decl`` (from the `Haskell 2010 Report`_), ::
+
+     decl → 'type' simpletype '=' type
+          → 'type' tyvar '=' type
+
+   and remove the production ``topdecl → 'type' simpletype '=' type`` from ``topdecl``.
+
+   Note that the second form allows a local binding for a lower-case ``tyvar``; these
+   synonyms may not be parameterized.
+
+#. The form ``decl → 'type' tyvar '=' type`` is not allowed at top-level.
+
+#. These new declaration forms introduce local type synonyms in terms, which scope over the same
+   region of code that other declarations in the same ``let`` / ``where`` clause scope over.
+
+   Like other type synonyms, local type synonyms may not be recursive.
+
+#. Wildcards are allowed in the right-hand side of local synonyms. At usage sites of the
+   synonym, the synonym is expanded. It is an error if that location does not allow wildcards.
+   The wildcard is understood to stand for just one type shared among all the expansions.
+
+Effects
+^^^^^^^
+
+1. We can now bind local type synonyms, avoiding the need to do so via pattern or result
+   signatures.
+
+#. One challenge is how to present these local synonyms in error messages. It might be
+   best to aggressively expand (unlike top-level type synonyms), especially because these
+   local synonyms might refer to other local type variables that are in scope. As we gain
+   experience with this new form, we can refine their appearance in error messages.
+
+#. Note that this proposal does *not* allow for top-level lower-case type synonyms. There
+   is nothing stopping us from doing so, but it would seem to violate expectations of Haskellers
+   and would be the first instance of a lower-case type variable being in scope at the top level.
+
+``let`` in patterns
+~~~~~~~~~~~~~~~~~~~
+
+This part of this proposal allows introducing a ``let``\ -binding in a pattern.
+The bound variable(s) scope over the same region of code as the pattern-bound
+variables do.
+
+The syntax for this feature is a bit awkward, and so this proposal presents
+two alternatives for discussion.
+
+Motivation
+^^^^^^^^^^
+
+1. A careful reader will note that allowing ``let`` for `type synonyms <#type-let>`_
+   does not, by itself, replace a binding such as that in ``f (True :: b) = ...``
+   because the current form binds the type variable in the pattern. This part
+   of the proposal closes this gap. See `examples <#let-in-pattern-example>`_ below.
```

Comment:
I don't understand this paragraph.

Reply:
It made sense to me at the time. I've removed it.